### ISSUE TITLE
Parsing anime files

### ISSUE THREAD

**Issue by lsrom** [tvrenamer/tvrenamer#117]
Hello again.

Parsing anime files as of right now isn't working. It is because of the different naming convention that uploaders use. Although sometimes filename is parsed, it is never correct as shown in example.  

`[AnimeRG] Dragon Ball Super - 047 [1080p] [x265] [pseudo].mkv`
`Dragon Ball Super - 059 Eng Sub x264.mkv`
`[PA]-Dragon-Ball-Z-001-[1080p][Hi8b][3849F7F5][V2]-(1).mkv`

Out of these three only the second one is parsed but still incorrectly. It is because the episode number 059 is parsed as s0e59 which obviously fails to load episode title.  

I think this can't be solved via regex as any changes to current regex will mess up normal parsing. My idea is new set of regexes or even classes to parse anime files. Because on some anime files parsing succeeds (even though incorrectly) it would be probably better to have **new button** in GUI something like **Parse As Anime** which would then start this new process.  

If anime parsing started only on files that previously couldn't be parsed it would mean that some could already be incorrectly parsed.  

Would you be ok with such changes? Or do you have something else in mind? I would implement it but I want to have permission for doing it as it might require significant changes to the code (new classes, changes in GUI).

**Comment by vipuldelwadia** [comment]
I think this issue is quite related to the earlier issue of handling shows with dates or many hundreds (or even thousands) of episode numbers, such as long running series or daily shows.

I agree that using more regexes won't solve it, however I wonder if we can have different modes (like you suggested) for the different types of shows (note that there might be some overlap in these types):

- regular
- anime
- daily
- season-less

I would be in favour of showing these different "modes" in the UI, and having tvrenamer auto pick the right mode based on what the file looks like (regex), but then the user can change the mode if it's wrong.

You can have a go at implementing it - I'm ok with new UI / changes in UI, as long as the simplicity is kept :)

**Comment by lsrom** [comment]
I agree with simplicity, that is important which is why I had another idea but you posted your answer before I could edit it.  

Instead of button for switching to long episode num parsing we could just run normal parsing + downloading episode names and only if downloading the episode name fails then call the new parsing procedure.  

This way no UI change is necessary, maybe just in preferences so users can turn this feature on and off.

**Comment by vipuldelwadia** [comment]
Yep that's another option, again you can have a play and see what works better :) I'm happy to try things out on a branch or pull request if you wanted to get some feedback or ideas.

**Comment by lsrom** [comment]
Just an update, I haven't forgotten about the project just don't have much time to work on plus I am currently stuck on problems with downloading episodes from The TVDB.  
It seems that `com.google.code.tvrenamer.model.showStore.getShow(String, ShowInformationListener)` returns incomplete listings.  
Have you ever encountered such a problem in the past? And if so, how did you solve it?

**Comment by vipuldelwadia** [comment]
Have you got an example of a show that returns incomplete listings?

**Comment by lsrom** [comment]
I am testing it with Dragon Ball Super. You can see tests I made in [my fork](https://github.com/lsrom/tvrenamer/blob/master/src/test/com/google/code/tvrenamer/TVRenamerSeasonlessTest.java). Note that the code is currently broken so it will probably not run but tests should work. If you run them you'll see that some of the episodes present in the TVDB output for Dragon Ball Super ([here](https://thetvdb.com/api/4A9560FF0B2670B2/series/295068/all/en.xml)) are missing and thus episodes have not been initialized.

I need to further investigate this if it isn't fault of my code as I am not yet fully comfortable with the code base because I work on it very sporadically. I'll continue work on it when I have the time.

**Comment by lsrom** [comment]
All problems with listings solved and all test are passing. `TVRenamer` and `TVRenamerSeasonless` now have the option to choose of they prefer DVD number or plain episode number. `TVRenamer` is set to prefer DVD number as before so no code is broken and everything works as before.  

`TVRenamerSeasonless` prefers by default normal episode number and this requires extra testing. If all anime shows are like DBS and normal episode numbering is the correct one, all we have to do is call `TVRenamerSeasonless.parseFilename()` in the right place and it's done.  

However if they differ in what correct number is, we need to introduce new algorithm for determining correct episode number, maybe even new listing provider.

Question: Should I make pull request with what I got now, which is working solution for parsing and renaming anime files which is used only in tests so everyone can take a look and maybe add to testing or should I fully implement it and we will test it in production?

**Comment by lsrom** [comment]
Pull request #125 added. TVRenamer is now able to use old or new parsing functions depending upon input filename.

**Comment by eprenamer** [comment]
I have a lot of ideas here.

I'm not sure what the value of the Season class really is, anyway.  The way TVDB has it organized, episodes are associated with a series, and that's it.  And while we don't necessarily have to follow them, and don't want to be overly tied to them, I think it makes a lot of sense.

That doesn't mean we would lose season data.  The "season number" is an attribute of the episode just like the episode number is, or the title, or the air date.  It just doesn't need to be an object.

Once it's organized this way internally, it opens things up much more for these kinds of episodes.  But then we have the parsing issues.  I think there's a lot we can do to expand it, there.

I'm currently taking a look at lsrom's PR, to see what he came up with.

**Comment by daveharris** [comment]
You make an interesting point @eprenamer, we don't do anything with a 'Season', it does really belong as an attribute of an episode.
I think I added it to model the 'real world' but if it's never used on it's own then we can easily get rid of it.
We have talked in the distant past about having the ability to have a kind 'retry parsing' function if it gets the show wrong, but that would be on a file but file basis so that wouldn't really work in this instance.

**Comment by eprenamer** [comment]
If the "real world" is the provider, then it definitely is not part of the current real world.

Possibly it was used by TVRage, and/or by other providers?

It does make things easier if our model matches the provider's, but it's certainly not necessary.  In this case, eliminating Season does both (matches the provider and makes things simpler)
